src/components/ConsoleInput.jsx

```diff
@@ -57,10 +73,41 @@ export default class ConsoleInput extends Component {
       editor.resize();
       editor.focus();
 
+      editor.commands.addCommand({
+        name: 'historyPrevious',
+        bindKey: 'Up',
+        exec: () => {
+          onPreviousConsoleHistory();
+        },
+      });
+
+      editor.commands.addCommand({
+        name: 'historyNext',
+        bindKey: 'Down',
+        exec: () => {
+          onNextConsoleHistory();
+        },
+      });
+
       session.on('change', ({action, lines}) => {
-        if (action === 'insert' && lines.length === 2) {
-          onInput(editor.getValue().replace('\n', ''));
-          editor.setValue('', 0);
+        const programmaticEdit = ['historyNext', 'historyPrevious'].includes(
+          editor.curOp.command.name,
+        );
+
+        if (programmaticEdit) {
+          return;
+        }
```

Comment:
I think we can actually drop this check—the check for `value !== this.props.currentInputValue` below seems to catch every case I can think of, doesn’t require specific knowledge of the kinds of things that might cause the input to change, and doesn’t require inspecting a transient state property on the editor. Unless there is a case I am missing?

Reply:
I agree that this isn't the most elegant solution, but it seems that due to the internal design of ace, this is the best approach.

The problem appears to me that when I have typed `x` in the console and hit the up arrow that will go to the last entry (let's call that `y`), in order to remove the x, ace will dispatch a `change` with action `remove` that will set the value to empty string, then another change event that will set the value to `y`. The first `remove` calls `componentDidUpdate` with empty string as the value, which is why the state input value would switch to empty string and back to the actual value whenever the user goes up and down in the history, in a previous edition of this PR.

Unfortunately the `remove` which sets the value to empty string is indistinguishable from the user deleting the input manually, unless we look at `editor.curOp.command.name` to see if it's happening during a history action.

This check isn't strictly necessary, since dispatching extra actions: x -> empty string -> y still gets to y, but especially while debugging having the input switch to empty string was annoying and with this check every user action can correspond to a single redux action.

It'd be great if calling `setValue` didn't also trigger `change`, but it does, so we have to prevent the update from triggering other updates using this check.